### PR TITLE
Add iproto requests for stream transactions

### LINKED ISSUE #2402: [2pt] ability to set timeout for iproto transactions was implemented
# Add `IPROTO_TIMEOUT 0x56` key

**Product:** Tarantool
**Root document:** https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/
**SME:** @ locker
**Peer reviewer:** @ patiencedaur 

# Details
A new `IPROTO_TIMEOUT 0x56` key has been added. Currently it is used to
set a timeout for transactions over iproto streams. It is stored in the
body of 'IPROTO_BEGIN' request. If user want's to specify timeout using
netbox (3s for example), he should use 'stream:begin({timeout = 3}).
Requested by @EvgenyMekhanik in https://github.com/tarantool/tarantool/commit/b9f7204b5e0d10b443c6f198e9f7f04e0d16a867.

### PR DESCRIPTION
**Pull request by p7nov**
Resolves #2402 #2922

### CODE REVIEW

File: doc/dev_guide/internals/box_protocol.rst [hunk 1]
```diff
@@ -1281,16 +1282,20 @@ each call to conn:new_stream() assigns a new number, starting with 1.
...
 The client makes stream transactions by sending, in order:
-IPROTO_BEGIN, the transaction data-change and query requests,
-IPROTO_COMMIT or IPROTO_ROLLBACK.
-Each request must contain the same IPROTO_STREAM_ID value.
-With streaming there is no need to add
-:ref:`IPROTO_FLAGS <box_protocol-flags>` and IPROTO_FLAG_COMMIT
-in the header of the last request of a transaction.
-Rollback will be automatic if disconnect occurs before commit is possible.
+
+1. IPROTO_BEGIN with an optional transaction timeout in the IPROTO_TIMEOUT field of the request body
```
Comment (locker) [inline doc/dev_guide/internals/box_protocol.rst]: It's unclear what the type of `IPROTO_TIMEOUT` is.

Overall, I don't like how IPROTO transactions are documented. We have a separate entry for each IPROTO command (`IPROTO_SELECT`, `IPROTO_REPLACE`, etc) with the description of header and data fields, but not for `IPROTO_BEGIN`, `IPROTO_COMMIT`, `IPROTO_ROLLBACK` - those commands are clustered together in one section, but they are no worse than other commands so why would we want to treat them in a different way? If I developed a connector to Tarantool, I'd prefer all IPROTO commands and their arguments to be listed explicitly.

[image](https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/)

Also, we already have entries for `IPROTO_COMMIT` and `IPROTO_ROLLBACK`, but those are actually `IPROTO_RAFT_COMMIT` and `IPROTO_RAFT_ROLLBACK` - see the picture above. This is confusing. Better rename them there.

File: doc/dev_guide/internals/box_protocol.rst [hunk 2]
```diff
@@ -1281,16 +1282,20 @@ each call to conn:new_stream() assigns a new number, starting with 1.
...
+2. the transaction data-change and query requests
+3. IPROTO_COMMIT or IPROTO_ROLLBACK
+
+All these requests must contain the same IPROTO_STREAM_ID value.
+
+With streaming, there is no need to add
+:ref:`IPROTO_FLAGS <box_protocol-flags>` and IPROTO_FLAG_COMMIT in the header
```
Comment (locker) [inline doc/dev_guide/internals/box_protocol.rst]: I understand that you just rearrange the text written before you, but `IPROTO_FLAG_COMMIT` has **absolutely nothing** to do with client transactions so saying that "there is no need to add `IPROTO_FLAG_COMMIT`" sounds confusing - like you may set it, but it's not necessary. In fact, those flags must not be used in client transactions at all - they will be silently ignored. `IPROTO_FLAG_COMMIT` is an internal flags that is used to store transactions in WAL and replicate them.

Comment (p7nov) [reply doc/dev_guide/internals/box_protocol.rst]: Got it, removing this sentence.

File: doc/dev_guide/internals/box_protocol.rst (hunk 1 above)
Comment (p7nov) [reply doc/dev_guide/internals/box_protocol.rst]: > Overall, I don't like how IPROTO transactions are documented

Yeah, I've noticed this inconsistency. Filed a new issue (#2922), will do in this PR together with proper description of IPROTO_TIMEOUT.

> Also, we already have entries for IPROTO_COMMIT and IPROTO_ROLLBACK, but those are actually IPROTO_RAFT_COMMIT and IPROTO_RAFT_ROLLBACK

Thanks, fixed.

File: doc/dev_guide/internals/box_protocol.rst [hunk 3]
```diff
@@ -706,6 +708,96 @@ IPROTO_SQL_TEXT (0x40) and statement-text (string) if executing an SQL string.
...
+
+.. // TODO: add link to transaction isolation docs once they're ready
+IPROTO_TXN_ISOLATION is the transaction isolation level. It can take
+the following values:
+
+.. // TODO: provide links to level descriptions
+- ``TXN_ISOLATION_DEFAULT = 0``	-- use the global default level
+- ``TXN_ISOLATION_READ_COMMITTED = 1`` -- read committed, but not confirmed changes
+- ``TXN_ISOLATION_READ_CONFIRMED = 2`` -- read only confirmed changes
+- ``TXN_ISOLATION_BEST_EFFORD = 3`` -- determine isolation level automatically
```
Comment (locker) [inline doc/dev_guide/internals/box_protocol.rst]: Typo: efford -> effort

**Review by locker — APPROVED** [review]

File: doc/dev_guide/internals/box_protocol.rst [hunk 4]
```diff
@@ -706,6 +708,96 @@ IPROTO_SQL_TEXT (0x40) and statement-text (string) if executing an SQL string.
...
+
+.. // TODO: add link to transaction isolation docs once they're ready
+IPROTO_TXN_ISOLATION is the transaction isolation level. It can take
+the following values:
+
+.. // TODO: provide links to level descriptions
+- ``TXN_ISOLATION_DEFAULT = 0``	-- use the global default level (default value)
+- ``TXN_ISOLATION_READ_COMMITTED = 1`` -- read changes that are committed but not confirmed yet
+- ``TXN_ISOLATION_READ_CONFIRMED = 2`` -- read confirmed changes
+- ``TXN_ISOLATION_BEST_EFFORT = 3`` -- determine isolation level automatically
```
Comment (patiencedaur) [inline doc/dev_guide/internals/box_protocol.rst:748]:
```suggestion
- ``TXN_ISOLATION_DEFAULT = 0``	-- use the global default level (default value).
- ``TXN_ISOLATION_READ_COMMITTED = 1`` -- read changes that are committed but not confirmed yet.
- ``TXN_ISOLATION_READ_CONFIRMED = 2`` -- read confirmed changes.
- ``TXN_ISOLATION_BEST_EFFORT = 3`` -- determine isolation level automatically.
```
You can find the list punctuation rules in our [style guide](https://www.tarantool.io/en/doc/latest/contributing/docs/style/#lists).

File: doc/dev_guide/internals/box_protocol.rst [hunk 5]
```diff
@@ -1281,16 +1373,18 @@ each call to conn:new_stream() assigns a new number, starting with 1.
...
-IPROTO_COMMIT or IPROTO_ROLLBACK.
-Each request must contain the same IPROTO_STREAM_ID value.
-With streaming there is no need to add
-:ref:`IPROTO_FLAGS <box_protocol-flags>` and IPROTO_FLAG_COMMIT
-in the header of the last request of a transaction.
-Rollback will be automatic if disconnect occurs before commit is possible.
+
+1. IPROTO_BEGIN with an optional transaction timeout in the IPROTO_TIMEOUT field of the request body
+2. the transaction data-change and query requests
+3. IPROTO_COMMIT or IPROTO_ROLLBACK
```
Comment (patiencedaur) [inline doc/dev_guide/internals/box_protocol.rst]:
```suggestion

1. IPROTO_BEGIN with an optional transaction timeout in the IPROTO_TIMEOUT field of the request body.
2. The transaction data-change and query requests.
3. IPROTO_COMMIT or IPROTO_ROLLBACK.
```

**Review by patiencedaur — APPROVED** [review]

**Pull request merged**